src/main/java/com/exactpro/th2/hand/services/HandBaseService.java

```diff
@@ -16,65 +16,135 @@
-		responseObserver.onNext(Empty.getDefaultInstance());
-		responseObserver.onCompleted();
+		try {
+			responseObserver.onNext(unregister(request));
+			responseObserver.onCompleted();
+		} catch (Exception e) {
+			LOGGER.error("Action failure, request: '{}'", toJson(request), e);
+			responseObserver.onError(e);
+		}
+
 	}
 
 	@Override
 	public void executeRhActionsBatch(RhActionsBatch request, StreamObserver<RhBatchResponse> responseObserver) {
-		logger.trace("Action: '{}', request: '{}'", "executeRhActionsBatch", TextFormat.shortDebugString(request));
-		responseObserver.onNext(messageHandler.handleActionsBatchRequest(request));
-		responseObserver.onCompleted();
+		LOGGER.trace("Action: 'executeRhActionsBatch', request: '{}'", toJson(request));
+        try {
+            responseObserver.onNext(executeRhActionsBatch(request));
+			responseObserver.onCompleted();
+        } catch (Exception e) {
+			LOGGER.error("Action failure, request: '{}'", toJson(request), e);
+            responseObserver.onError(e);
+        }
 	}
 
 	@Override
 	public void dispose() {
 		try {
 			this.messageHandler.close();
 		} catch (Exception e) {
-			logger.error("Error while disposing message handler", e);
+			LOGGER.error("Error while disposing message handler", e);
 		}
 	}
+
+	/**
+	 * @throws RhConfigurationException
+	 */
+	@SuppressWarnings("JavadocDeclaration")
+	@Override
+	public RhSessionID register(RhTargetServer targetServer) {
+        try {
+			String sessionId = messageHandler.getRhConnectionManager().createSessionHandler(targetServer.getTarget()).getId();
+			return RhSessionID.newBuilder().setId(sessionId).setSessionAlias(messageHandler.getConfig().getSessionAlias()).build();
+		} catch (RhConfigurationException e) {
+			sneakyThrow(e);
+			return null;
+		}
+	}
+
+	@Override
+	public RhSessionID register(RhTargetServer input, Map<String, String> properties) {
+		return register(input);
+	}
+
+	@Override
+	public Empty unregister(RhSessionID request) {
+		messageHandler.getRhConnectionManager().closeSessionHandler(request.getId());
+		return Empty.getDefaultInstance();
+	}
+
+	@Override
+	public Empty unregister(RhSessionID input, Map<String, String> properties) {
+		return unregister(input);
+	}
+
+	/**
+	 * @throws IOException
+	 */
+	@SuppressWarnings("JavadocDeclaration")
+	@Override
+	public RhBatchResponse executeRhActionsBatch(RhActionsBatch request) {
+        try {
+            return messageHandler.handleActionsBatchRequest(request);
+        } catch (IOException e) {
+			sneakyThrow(e);
+			return null;
+        }
+    }
+
+	@Override
+	public RhBatchResponse executeRhActionsBatch(RhActionsBatch input, Map<String, String> properties) {
+		return executeRhActionsBatch(input);
+	}
+
+	// This hack is used to avoid wrapping cause error
+	private static <E extends Throwable> void sneakyThrow(Throwable e) throws E {
+		//noinspection unchecked
+		throw (E) e;
+	}
```

Comment:
And why would we use this hack in java? How will anybody know that this method might throw a checked exception?
Even Kotlin provides annotations to help Java understand if a checked exception might be thrown by the method. But we use this hack to hide this information. Why don't we just wrap the exception in runtime exception in this case?

Reply:
I think you are right because gRPC StatusRuntimeException isn't handled in act-win-core